bindings/profilers/wall.cc

```diff
@@ -25,213 +26,455 @@
 #include "../per-isolate-data.hh"
 #include "wall.hh"
 
+#ifndef _WIN32
+#define DD_WALL_USE_SIGPROF
+#else
+#undef DD_WALL_USE_SIGPROF
+#endif
+
 using namespace v8;
 
 namespace dd {
 
-Local<Object> CreateTimeNode(Local<String> name,
-                             Local<String> scriptName,
-                             Local<Integer> scriptId,
-                             Local<Integer> lineNumber,
-                             Local<Integer> columnNumber,
-                             Local<Integer> hitCount,
-                             Local<Array> children) {
-  Local<Object> js_node = Nan::New<Object>();
-  Nan::Set(js_node, Nan::New<String>("name").ToLocalChecked(), name);
-  Nan::Set(
-      js_node, Nan::New<String>("scriptName").ToLocalChecked(), scriptName);
-  Nan::Set(js_node, Nan::New<String>("scriptId").ToLocalChecked(), scriptId);
-  Nan::Set(
-      js_node, Nan::New<String>("lineNumber").ToLocalChecked(), lineNumber);
-  Nan::Set(
-      js_node, Nan::New<String>("columnNumber").ToLocalChecked(), columnNumber);
-  Nan::Set(js_node, Nan::New<String>("hitCount").ToLocalChecked(), hitCount);
-  Nan::Set(js_node, Nan::New<String>("children").ToLocalChecked(), children);
-
-  return js_node;
+using ProfilerMap = std::unordered_map<Isolate*, WallProfiler*>;
+
+static std::atomic<ProfilerMap*> profilers(new ProfilerMap());
+
+#ifdef DD_WALL_USE_SIGPROF
+static void (*old_handler)(int, siginfo_t*, void*) = nullptr;
+
+static void sighandler(int sig, siginfo_t* info, void* context) {
+  auto isolate = Isolate::GetCurrent();
+  WallProfiler* prof = nullptr;
+  auto prof_map = profilers.exchange(nullptr);
```

Comment:
Very elegant solution !
This way sighandler will always be able to "lock" the map without using a mutex, nor using a CAS loop.
All the complexity / latency is moved to updateProfilers, but this is not a problem since it is rarely called.